src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1245 @@
+/*-------------------------------------------------------------------------
+ *
+ * shard_pruning.c
+ *   Shard pruning related code.
+ *
+ * Shard pruning works in two phases: First it analyses the expressions that
+ * are pruned upon, and builds a list of ORed constraints for pruning. This
+ * happens in PrunableExpressions().  The reason for keeping this separate
+ * from actually pruning away shards is that it has to only be done once for
+ * all shards.
+ *
+ * In the second phase, the shards are pruned using the constraints built in
+ * the first phase.  Each constraint (a set of ANDed constraints) is evaluated
+ * by PruneOne(), and the results of all results is then ORed/unioned
+ * together.  Depending on the types of constraints found, and the type of
+ * partitioning used, different methods of pruning are employed.  Ranging from
```

Comment:
It would be nice if this documentation was a bit more explicit about what we're doing.

The goal of shard pruning is to find a minimal (super)set of shards that need to be queried to find rows matching the expression in a query.

In `PruneShards`, we first compute a simplified disjunctive normal form (DNF) of the expression as a list of pruning instances. Each pruning instance contains all AND-ed constraints on the partition column. An OR expression will result in two or more new pruning instances for the subexpression.

We use the distributive property for constraints of the form P AND (Q OR R) to rewrite it to (P AND Q) OR (P AND R). However, we do not distribute nested expressions. While (P OR Q) AND (R OR S) is logically equivalent to (P AND R) OR (P AND S) OR (Q AND R) OR (Q AND S), in our implementation it becomes P OR Q OR R OR S. This is acceptable since this will always result in a superset.

We then evaluate each pruning instance in the disjunction through the following steps:

1. If there is a constant equality constraint on the partition column, find the shard interval in which the constant falls
2. If there is a hash range constraint on the partition column, find the shard interval matching the range
3. If there are range constraints on the partition column, find the shard intervals that overlap with the range
4. If there are overlapping shards, exhaustively search all shards that are not excluded by constraints

Finally, we take the union of the shards found by each pruning instance.